Allow adding playlist items to beatmap collections via context menu

frenzibyte opened the pull request:
- Addresses https://github.com/ppy/osu/discussions/19019

Will look into adding test coverage a bit later if deemed necessary, need to figure out how to create custom collections in tests without messing up the main databases.

---

https://user-images.githubusercontent.com/22781491/179123652-440fa82e-7f4f-4d81-8a30-b1cc024de45a.mp4

frenzibyte committed:
Split collection toggle menu item to own class

frenzibyte committed:
Add collection context menu to room playlist items

frenzibyte committed:
Cover online-play room screens with context menu containers

frenzibyte committed:
Remove no longer required context menu container in `ParticipantsList`

frenzibyte committed:
Add missing license header

smoogipoo commented on osu.Game/Screens/OnlinePlay/DrawableRoomPlaylistItem.cs:
```diff
@@ -470,6 +490,30 @@ protected override bool OnClick(ClickEvent e)
+                        items.Add(new OsuMenuItem("Collections") { Items = collectionItems });
+                    }
+                    else
+                        items.Add(new OsuMenuItem("Download to add to collection") { Action = { Disabled = true } });
```
I really don't think this is required. It's just adding to the complexity of this whole thing without (imo) providing much use even if it did work.

peppy replied:
Instead of this, why not add the standard "Details..." to this context menu? I'd really hope that this becomes a standard context menu applied to any beatmap context going forward.

https://github.com/ppy/osu/blob/f8830c6850128456266c82de83273204f8b74ac0/osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs#L223-L224

frenzibyte committed:
Merge branch 'master' into drawable-playlist-item-collection

frenzibyte committed:
Centralise creation of playlist in test scene

frenzibyte committed:
Handle `GetBeatmapSetRequest` on test room requests handler

Required for `BeatmapSetOverlay` lookups to work under dummy API. Not
100% sure about it, but works for now.

frenzibyte committed:
Add beatmap details menu item to playlist items

frenzibyte committed:
Add test coverage for details item

frenzibyte committed:
Add test coverage for collection items

frenzibyte commented:
Added test coverage (went through hell with beatmap lookup issues for that) and addressed mentioned concern.

frenzibyte committed:
Fix potential nullref in `ContextMenuItems`

frenzibyte commented:
Quite lost on how the [test failure from above](https://github.com/ppy/osu/runs/7331747090?check_suite_focus=true) occurred, tempted to remove it at this point...

frenzibyte committed:
Fix `CollectionManager` opening file multiple times across test scene

peppy commented:
I don't quite get the last commit, but it also doesn't seem to fix the issue (same test failure happened).

frenzibyte commented:
The last commit was for fixing [this test failure](https://github.com/ppy/osu/runs/7332428838?check_suite_focus=true), but I can't say I know what's exactly causing it, only following what's already done in other places (creating collection manager on BDL rather than per test case).

peppy commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:
```diff
@@ -81,134 +82,138 @@ protected override void LoadComplete()
             RelativeSizeAxes = Axes.Both,
             Padding = new MarginPadding { Horizontal = 5, Vertical = 10 },
-            Child = new GridContainer
+            Child = new OsuContextMenuContainer
```
Heavily unsure about putting these at a screen level. They are usually nested inside of scroll views so they correctly dismiss context menus when the item displaying them scrolls into a masked away state.

frenzibyte replied:
I've read through `ContextMenuContainer` and I'm not sure context menus dismiss at all when items are masked away? It only happens when they are no longer present (perhaps an oversight?)

In addition, I've tested using `ParticipantsList` with three different ways of placing the context menu container just to figure out what seems more correct:

 - Placed outside `ParticipantsList`'s scroll content and filling entire screen (current code in PR):
   
   https://user-images.githubusercontent.com/22781491/179121234-cc63438d-afb8-401a-aed1-531852fafa78.mp4

 - Placed outside `ParticipantsList`'s scroll content and filling to the list size (master):
 
   https://user-images.githubusercontent.com/22781491/179121280-5d8b7ca7-10be-41d1-bf21-0b0d3b107061.mp4

 - Placed inside `ParticipantsList`'s scroll content:
   
   https://user-images.githubusercontent.com/22781491/179121342-53afabf1-c61a-4fb1-a4ed-0cf4e84d03e2.mp4

Posting this since it looks more correct to place outside the list and filling entire playlists/multiplayer screen, so that the context menu still displays at the cursor's position when right-clicking the last displayable item.

peppy commented:
I think you're testing wrong here. You're testing *CollectionManager* logic in a *context menu* test. This would be like testing that a beatmap imports correctly from a song select carousel test. I think it's unnecessary and should be simplified. At most I'd check that the items for collections from the database are created in the menu, but even then I'd hope that is using a mock collection manager and not dealing with `collection.db` files on disk.

Keep it simple and avoid spending hours fixing silly tests we don't need in the first place.

peppy commented:
For this particular change, I'd probably be fine with no test coverage, FWIW.

frenzibyte committed:
Remove unnecessary test coverage

RIP hours.

frenzibyte commented:
Just to note, my intentions with the test coverage was to ensure the context menu items behave correctly in case they regress later due to changes in the `DrawableRoomPlaylistItem` structure (ensuring "Details" works, ensuring "Collections" shows collections and clicking on one of them adds to collection and updates item state, etc.), but I'm already giving up on making this test work without failing every 2 attempts.

I've removed all test coverage, only kept minor changes to the test scene and a drive-by fix to `TestDownloadButtonHiddenWhenBeatmapExists` (doesn't work when running a second time on interactive/visual tests).

peppy commented on osu.Game/Collections/CollectionToggleMenuItem.cs:
```diff
@@ -0,0 +1,23 @@
+        public CollectionToggleMenuItem(BeatmapCollection collection, IBeatmapInfo beatmap)
+            : base(collection.Name.Value, MenuItemType.Standard, s =>
+            {
+                if (s)
```
rename this to anything but `s`

frenzibyte replied:
Copy-pasted from previous place (in carousel beatmap class), but do agree with that.

peppy commented on osu.Game/Screens/OnlinePlay/DrawableRoomPlaylistItem.cs:
```diff
@@ -304,6 +321,15 @@ private void refresh()
             difficultyIconContainer.FadeInFromZero(500, Easing.OutQuint);
             mainFillFlow.FadeInFromZero(500, Easing.OutQuint);
+
+            downloadTracker?.RemoveAndDisposeImmediately();
+
+            if (beatmap != null)
+            {
+                Debug.Assert(beatmap.BeatmapSet != null);
+                downloadTracker = new BeatmapDownloadTracker(beatmap.BeatmapSet);
+                AddInternal(downloadTracker);
```
This is too much overhead just to maybe display context menu items at some point.

I dislike download trackers being everywhere and don't want ones like this that are probably never even queried. Just use a `BeatmapManager` query at the point of constructing the menu items.

frenzibyte replied:
Had considered querying `BeatmapManager`, but thought it wouldn't hurt to add a download tracker (well, maybe bad because there's literally a download button next to the tracker, but yeah).

frenzibyte committed:
Use proper variable name

frenzibyte committed:
Replace download tracker with local querying

peppy reviewed (approved):
Haven't tested, seems okay.
